### PR TITLE
Patch coalesce

### PR DESCRIPTION
**Pull request by webmiche**
The first patch for coalesce. This only applies the redundant heuristic, i.e., it checks for sets that are contained in others and deletes those.

I would especially like to get feedback on the new looping technique. Currently, I mark `FlatAffineConstraints` as being eliminated and then stop considering them.

### CODE REVIEW

**Comment by webmiche** [comment]
For the current version, I decided to return a PresburgerSet and not change the PresburgerSet `coalesce` is called on. I think there is a discussion to be had, if we want this to happen in-place or not. As it currently was not done in-place and the test depend on this fact, I left it like that.

File: mlir/include/mlir/Analysis/PresburgerSet.h [hunk 1]
```diff
@@ -90,6 +90,11 @@ class PresburgerSet {
   /// any of the FACs in the union are unbounded.
   bool findIntegerSample(SmallVectorImpl<int64_t> &sample);
 
+  /// eliminates FlatAffineConstraints, that are fully contained within other
```
Comment (superty) [inline mlir/include/mlir/Analysis/PresburgerSet.h]: I think this could be written better. e.g.

```quote
Simplifies the representation of the PresburgerSet.

In particular, removes all FACs which are subsets of other FACs in the union.
```

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 2]
```diff
@@ -234,6 +234,9 @@ class Simplex {
   void print(raw_ostream &os) const;
   void dump() const;
 
+  /// Check if the given constraint is redundant
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: Although it's clear from the function signature, it would be nice to specify in the doc that we refer to a new constraint rather than an existing constraint. e.g. "Check if the specified inequality already holds in the polytope."

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 3]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+      bool contained = true;
+      for (unsigned i = 0; i < numIneq; ++i) {
+        contained &= simplex.isRedundant(bs1.getInequality(i));
+      }
+      for (unsigned i = 0; i < numEq; ++i) {
+        contained &= simplex.isRedundant(bs1.getEquality(i));
+      }
+
+      if (contained) {
+        marked[j] = true;
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Here and below, you can skip the braces for single-statement ifs

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 4]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+      }
+    }
+  }
+
+  for (unsigned i = 0; i < basicSetVector.size(); ++i) {
+    if (!marked[i]) {
+      newSet.unionFACInPlace(basicSetVector[i]);
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Skip braces for single statement ifs.

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 5]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
   return false;
 }
 
+PresburgerSet PresburgerSet::coalesce() {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(this->getNumDims(), this->getNumSyms());
+  ArrayRef<FlatAffineConstraints> basicSetVector =
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: I don't understand why this is needed. Just use `flatAffineConstraints` directly?

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 6]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
+      Simplex simplex(bs2);
+
+      bool contained = true;
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: I think you can factor this out into a separate function, probably `Simplex::contains(const FlatAffineConstraints &fac)`. You can then also return early when any constraint turns out to be non-redundant. (I'm suggesting making it a member function of Simplex rather than FlatAffineConstraints so that you can re-use the same Simplex object many times as suggested above.)

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 7]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
   return false;
 }
 
+PresburgerSet PresburgerSet::coalesce() {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: I think the same code should work with sets with symbols, so this shouldn't be necessary.

File: mlir/unittests/Analysis/Presburger/SimplexTest.cpp [hunk 8]
```diff
@@ -383,4 +383,17 @@ TEST(SimplexTest, addInequality_already_redundant) {
...
+  simplex.addEquality({-1, 1, 0});   // [2]: y = x.
+
+  EXPECT_TRUE(simplex.isRedundant({-1, 0, 2})); // x <= 2.
+  EXPECT_TRUE(simplex.isRedundant({0, 1, 0}));  // y >= 0.
+
+  EXPECT_FALSE(simplex.isRedundant({-1, 0, -1})); // x <= -1.
+  EXPECT_FALSE(simplex.isRedundant({0, 1, -2}));  // y >= 2.
+}
+
 } // namespace mlir
```
Comment (superty) [inline mlir/unittests/Analysis/Presburger/SimplexTest.cpp:485]: Can you add a test where the new inequality doesn't make the set empty?

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 9]
```diff
@@ -1195,4 +1195,9 @@ void Simplex::print(raw_ostream &os) const {
 
 void Simplex::dump() const { print(llvm::errs()); }
 
+bool Simplex::isRedundant(ArrayRef<int64_t> coeffs) {
+  Optional<Fraction> minimum = computeOptimum(Direction::Down, coeffs);
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Simplex.cpp:1262]: Doesn't this assert-fail when the new inequality makes the set empty?

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 10]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
+    if (marked[i])
+      continue;
+    FlatAffineConstraints bs1 = basicSetVector[i];
+
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Please use `for (unsigned j = 0, e = basicSetVector.size(); j < e; ++j) {` form.

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 11]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
 
+PresburgerSet PresburgerSet::coalesce() {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(this->getNumDims(), this->getNumSyms());
+  ArrayRef<FlatAffineConstraints> basicSetVector =
+      flatAffineConstraints;
+  SmallVector<bool, 4> marked(this->getNumFACs());
+
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Here and below, please avoid calling a function in the loop condition unless necessary, e.g., `for (unsigned i = 0, e = basicSetVector.size(); i < e; ++i) {`. Also, use, preincrements.

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 12]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+    FlatAffineConstraints bs1 = basicSetVector[i];
+
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
+      Simplex simplex(bs2);
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: I think you should flip the loop and construct the Simplex object in the outer loop. Then you only need construct it once per basic set instead of one per pair of basic sets. (worst case)

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 13]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(this->getNumDims(), this->getNumSyms());
+  ArrayRef<FlatAffineConstraints> basicSetVector =
+      flatAffineConstraints;
+  SmallVector<bool, 4> marked(this->getNumFACs());
+
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
+    if (marked[i])
+      continue;
+    FlatAffineConstraints bs1 = basicSetVector[i];
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Is this copy necessary?

File: mlir/lib/Analysis/PresburgerSet.cpp [hunk 14]
```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
...
+
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
```
Comment (superty) [inline mlir/lib/Analysis/PresburgerSet.cpp]: Is this copy necessary?

File: mlir/unittests/Analysis/PresburgerSetTest.cpp [hunk 15]
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
   EXPECT_FALSE(rect.complement().isEqual(square.complement()));
 }
 
+void expectCoalesce(size_t expectedNumBasicSets, PresburgerSet set) {
+  PresburgerSet newSet = set.coalesce();
+  set.dump();
```
Comment (superty) [inline mlir/unittests/Analysis/PresburgerSetTest.cpp]: We probably don't want to dump output for passing cases. If you want to keep this, I would suggest printing it only for failing cases.

**Review by superty — COMMENTED** [review]
I think returning a new set is fine for now. Re: looping strategy, I left some inline comments.

File: mlir/lib/Analysis/PresburgerSet.cpp (hunk 3 above)
Comment (webmiche) [reply mlir/lib/Analysis/PresburgerSet.cpp]: Is this also usually done for single statement fors? (5 lines earlier or so)

Comment (superty) [reply mlir/lib/Analysis/PresburgerSet.cpp]: Yes

File: mlir/lib/Analysis/PresburgerSet.cpp (hunk 6 above)
Comment (webmiche) [reply mlir/lib/Analysis/PresburgerSet.cpp]: Done.
I decided to name the function `Simplex::containedIn(const FlatAffineConstraints &fac)` as it actually checks whether the polytope behind the simplex is `contained in` the specified FAC.

File: mlir/unittests/Analysis/Presburger/SimplexTest.cpp (hunk 8 above)
Comment (webmiche) [reply mlir/unittests/Analysis/Presburger/SimplexTest.cpp:485]: I was not 100% sure what you meant here. I decided to change the test polytope x >= 0, y <= 2, y = x and added the check for y >= 1, which is, in this case, a cut inequality and not redundant while not making the set empty.

File: mlir/lib/Analysis/Presburger/Simplex.cpp (hunk 9 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:1262]: I am also not 100% sure what you mean here, I guess it relates to the similar comment above.

Comment (superty) [reply mlir/lib/Analysis/Presburger/Simplex.cpp:1262]: Nevermind, I was confused.

File: mlir/unittests/Analysis/PresburgerSetTest.cpp [hunk 16]
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
...
+  newSet.dump();
+  EXPECT_TRUE(set.isEqual(newSet));
+  EXPECT_TRUE(expectedNumBasicSets == newSet.getNumFACs());
+}
+
+TEST(SetTest, coalescecontainedOneDim) {
```
Comment (superty) [inline mlir/unittests/Analysis/PresburgerSetTest.cpp]: Please change it to coalesceContained, coalesceSeparate

File: mlir/unittests/Analysis/PresburgerSetTest.cpp [hunk 17]
```diff
@@ -591,4 +591,64 @@ TEST(SetTest, isEqual) {
...
+                                                  {0, 1, 0},    // y >= 0
+                                                  {0, -1, 3}}), // y <= 3.
+                             makeFACFromIneqs(2, {{1, 0, 0},    // x >= 0.
+                                                  {-1, 0, 3},   // x <= 3.
+                                                  {0, 1, -2},   // y >= 2.
+                                                  {0, -1, 3}}), // y <= 3
+                         });
+  expectCoalesce(1, set);
+}
+
```
Comment (superty) [inline mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: Please also add intersecting cases (some constraints are redundant but some are not) a case where one set is empty and the other is not, and case where both sets are empty.

Comment (superty) [reply mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: And also case where one is universe, both are universe, one is universe and one is empty

Comment (webmiche) [reply mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: Added.
You were right, when the first set is empty, the current implementation crashed. I took care of that by checking whether it is and, if so, marking it to be removed and continuing.
When both sets are empty, the current implementation returns a set without FACs in the FACvector. Is this how we represent empty sets? :)

Comment (superty) [reply mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: Yes, the union of nothing is considered to be empty (since the identity for union is the empty set)

Comment (superty) [reply mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: > I took care of that by checking whether it is and, if so, marking it to be removed and continuing.

The contains thing should support empty FACs. Please also add unit tests for it

Comment (webmiche) [reply mlir/unittests/Analysis/PresburgerSetTest.cpp:787]: Added a ton of additional tests

File: mlir/unittests/Analysis/Presburger/SimplexTest.cpp [hunk 18]
```diff
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "mlir/Analysis/Presburger/Simplex.h"
+#include "../utils.h"
```
Comment (Groverkss) [inline mlir/unittests/Analysis/Presburger/SimplexTest.cpp]: Can this path be made absolute?

File: mlir/unittests/Analysis/PresburgerSetTest.cpp [hunk 19]
```diff
@@ -15,6 +15,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "mlir/Analysis/PresburgerSet.h"
+#include "./utils.h"
```
Comment (Groverkss) [inline mlir/unittests/Analysis/PresburgerSetTest.cpp]: Can this path be made absolute?

File: mlir/unittests/Analysis/utils.h [hunk 20]
```diff
@@ -0,0 +1,41 @@
+//===- utils.h - utility functions for tests ------------------------------===//
```
Comment (Groverkss) [inline mlir/unittests/Analysis/utils.h]: I think all file names in the MLIR codebase follow UpperCamelCase. It would be better to rename this file to "FACUtils.h" or "Utils.h".

File: mlir/unittests/Analysis/Presburger/SimplexTest.cpp (hunk 18 above)
Comment (webmiche) [reply mlir/unittests/Analysis/Presburger/SimplexTest.cpp]: I would have to move this into the include directory. I am not certain how these kinds of things are usually done, but if you feel having them in the includes makes more sense, I can certainly do that.
